plugins/producer_plugin/producer_plugin.cpp

```diff
@@ -2909,147 +2862,143 @@ void producer_plugin_impl::start_write_window() {
+// Reschedule any exhausted read-only transactions from the last block
+void producer_plugin_impl::repost_exhausted_transactions(const fc::time_point& deadline) {
+   if ( !_ro_exhausted_trx_queue.empty() ) {
+      // post any exhausted back into read_only queue with slightly higher priority (low+1) so they are executed first
+      ro_trx_t t;
+      while( fc::time_point::now() < deadline && _ro_exhausted_trx_queue.pop_front(t) ) {
+         app().executor().post(priority::low+1, exec_queue::read_only, [this, trx{std::move(t.trx)}, next{std::move(t.next)}]() mutable {
+            push_read_only_transaction( std::move(trx), std::move(next) );
+         } );
+      }
+   }
 }
 
-// Called from a read_only_trx execution thread
+// Called from a read_only_trx execution thread, or from app thread when executing exclusively
 // Return whether the trx needs to be retried in next read window
-bool producer_plugin_impl::push_read_only_transaction(
-            const transaction_metadata_ptr& trx,
-            next_function<transaction_trace_ptr> next,
-            const fc::time_point& read_window_start_time) {
+bool producer_plugin_impl::push_read_only_transaction(transaction_metadata_ptr trx, next_function<transaction_trace_ptr> next) {
    auto retry = false;
-   chain::controller& chain = chain_plug->chain();
-
-   if( !chain.is_building_block() ) {
-      // try next round
-      return true;
-   }
 
    try {
-      const auto block_deadline = calculate_block_deadline( chain.pending_block_time() );
       auto start = fc::time_point::now();
-      auto time_used_in_read_window_us = ( start - read_window_start_time );
-      if ( time_used_in_read_window_us >= _ro_read_window_time_us ) {
-         // already passed read-window deadline, try next round
+      chain::controller& chain = chain_plug->chain();
+      if ( !chain.is_building_block() ) {
+         _ro_exhausted_trx_queue.push_front( {std::move(trx), std::move(next)} );
          return true;
       }
 
-      auto remaining_time_in_read_window_us = _ro_read_window_time_us - time_used_in_read_window_us;
-      // Ensure the trx to finish by the end of read-window.
-      auto window_deadline = std::min( start + remaining_time_in_read_window_us, block_deadline );
-
-      auto trace = chain.push_transaction( trx, window_deadline, _ro_max_trx_time_us, 0, false, 0 );
-      if ( _log.is_enabled( fc::log_level::debug ) ) {
-         auto dur = fc::time_point::now() - start;
-         _ro_all_threads_exec_time_us += dur.count();
+      // when executing on the main thread while in the write window, need to switch db mode to read only
+      // _ro_in_read_only_mode can only be false if running on main thread as it is only modified from the main thread
+      auto db_read_only_mode_guard = fc::make_scoped_exit([&]{
+         if( !_ro_in_read_only_mode )
+            chain.unset_db_read_only_mode();
+      });
+      if( !_ro_in_read_only_mode ) {
+         chain.set_db_read_only_mode();
```

Comment:
Why do we need to set and unset the `db_read_only_mode` when we are not in read only mode (so executing on a single thread).

Reply:
Because they are interleaved with non-read-only transactions. We set the db read only mode so we enforce read-only behavior for read-only transactions during the execution of the read-only transaction.

Reply:
I talked with Kevin and I understand now why we want to fail the transaction even in single thread mode (where it could execute successfully), because a user should have non readonly transactions submitted as readonly fail consistently.